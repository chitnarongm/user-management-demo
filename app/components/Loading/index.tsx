"use client";
import { Box, CircularProgress, styled } from "@mui/material";
import { FunctionComponent } from "react";

const LoadingWrapper = styled(Box)({
  position: "absolute",
  display: "flex",
  height: "100%",
  width: "100%",
  alignItems: "center",
  justifyContent: "center",
  top: 0,
  zIndex: 9999,
  backgroundColor: "rgba(255, 255, 255, 0.7)",
});

const Loading: FunctionComponent = () => {
  return (
    <LoadingWrapper>
      <CircularProgress data-testid="loading" />
    </LoadingWrapper>
  );
};

export default Loading;
