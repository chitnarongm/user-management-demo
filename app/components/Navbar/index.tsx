"use client";
import LogOutIcon from "@mui/icons-material/Logout";
import { AppBar, Box, Button, Toolbar, Typography } from "@mui/material";
import { signOut, useSession } from "next-auth/react";
import { FunctionComponent } from "react";

const Navbar: FunctionComponent = () => {
  const { data: session } = useSession();

  const handleLogOut = () => {
    signOut({ callbackUrl: "/login" });
  };

  return (
    <Box data-testid="navbar" sx={{ maxHeight: "64px", width: "100%", position: "fixed", top: 0, zIndex: "1100" }}>
      <AppBar position="static">
        <Toolbar>
          <Typography data-testid="navbar-title" variant="h5" component="div" sx={{ flexGrow: 1 }}>
            User Management
          </Typography>
          {session && (
            <Box display="flex" alignItems="center">
              <Box marginRight="24px">
                <Typography data-testid="user-info" lineHeight={0.75} fontWeight="bold" variant="subtitle2">
                  {session.user?.email}
                </Typography>
              </Box>
              <Button data-testid="logout-button" onClick={handleLogOut} color="inherit" startIcon={<LogOutIcon />}>
                Log Out
              </Button>
            </Box>
          )}
        </Toolbar>
      </AppBar>
    </Box>
  );
};

export default Navbar;
