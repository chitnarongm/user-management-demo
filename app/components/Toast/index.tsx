"use client";
import styled from "@emotion/styled";
import CloseIcon from "@mui/icons-material/Close";
import { Box } from "@mui/material";
import { closeSnackbar, enqueueSnackbar, OptionsObject, SnackbarKey } from "notistack";

const CloseButton = styled(CloseIcon)({
  cursor: "pointer",
});

const action = (snackbarId: SnackbarKey) => {
  return (
    <Box role="toast-alert-close-button" onClick={() => closeSnackbar(snackbarId)}>
      <CloseButton />
    </Box>
  );
};

const options: OptionsObject = {
  variant: "error",
  action,
  autoHideDuration: 3000,
};

export const openToast = (error: any) => {
  enqueueSnackbar(<Box role="toast-alert">{error}</Box>, options);
};
