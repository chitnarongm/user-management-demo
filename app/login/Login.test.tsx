import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { SignInResponse, signIn } from "next-auth/react";
import { useRouter } from "next/navigation";
import { openToast } from "../components/Toast";
import LogIn from "./page";

jest.mock("next-auth/react");
jest.mock("next/navigation");
jest.mock("../components/Toast");

describe("Login", () => {
  const renderLogin = () => {
    return render(<LogIn />);
  };

  it("should render correctly", () => {
    const { asFragment } = renderLogin();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should display information correctly", () => {
    renderLogin();

    expect(screen.getByTestId("login-page-title")).toHaveTextContent("Log In");

    expect(screen.getByTestId("login-email-label")).toHaveTextContent("Email");
    expect(screen.getByTestId("login-email-input")).toHaveDisplayValue("");

    expect(screen.getByTestId("login-password-label")).toHaveTextContent("Password");
    expect(screen.getByTestId("login-password-input")).toHaveDisplayValue("");

    expect(screen.getByTestId("login-submit-button")).toHaveTextContent("Log In");
  });

  describe("Email Field", () => {
    it("should validate required error correctly", async () => {
      renderLogin();

      await userEvent.click(screen.getByTestId("login-submit-button"));

      expect(screen.getByTestId("login-email-error")).toHaveTextContent("email is a required field");
    });

    it("should validate email pattern correctly", async () => {
      renderLogin();

      await userEvent.type(screen.getByTestId("login-email-input"), "invalid..email..pattern");

      await userEvent.click(screen.getByTestId("login-submit-button"));

      expect(screen.getByTestId("login-email-error")).toHaveTextContent("Invalid Email Format.");
    });
  });

  describe("Password Field", () => {
    it("should validate required error correctly", async () => {
      renderLogin();

      await userEvent.click(screen.getByTestId("login-submit-button"));

      expect(screen.getByTestId("login-password-error")).toHaveTextContent("password is a required field");
    });

    it("should validate password length correctly", async () => {
      renderLogin();

      await userEvent.type(screen.getByTestId("login-password-input"), "12345");

      await userEvent.click(screen.getByTestId("login-submit-button"));

      expect(screen.getByTestId("login-password-error")).toHaveTextContent("Password length must be more than 6.");
    });

    it("should display password visibility correctly", async () => {
      renderLogin();

      expect(screen.getByTestId("login-password-input")).toHaveAttribute("type", "password");

      await userEvent.click(screen.getByTestId("password-visibility-icon"));

      expect(screen.getByTestId("login-password-input")).toHaveAttribute("type", "text");

      await userEvent.click(screen.getByTestId("password-visibility-icon"));

      expect(screen.getByTestId("login-password-input")).toHaveAttribute("type", "password");
    });
  });

  describe("Login Action", () => {
    const replaceMock = jest.fn();

    beforeEach(() => {
      jest.mocked(useRouter).mockReturnValue({ replace: replaceMock } as any);
      jest.mocked(signIn).mockResolvedValue({ ok: true } as SignInResponse);
    });

    it("should redirect to /user-management page when login success", async () => {
      renderLogin();

      await userEvent.type(screen.getByTestId("login-email-input"), "mail@mail.com");
      await userEvent.type(screen.getByTestId("login-password-input"), "123456");

      await userEvent.click(screen.getByTestId("login-submit-button"));

      expect(replaceMock).toHaveBeenCalledWith("/user-management");
      expect(openToast).not.toHaveBeenCalled();
    });

    it("should still be in login page and display toast error when login failed with status 401", async () => {
      jest.mocked(signIn).mockRejectedValue({ status: 401 } as SignInResponse);

      renderLogin();

      await userEvent.type(screen.getByTestId("login-email-input"), "mail@mail.com");
      await userEvent.type(screen.getByTestId("login-password-input"), "123456");

      await userEvent.click(screen.getByTestId("login-submit-button"));

      expect(replaceMock).not.toHaveBeenCalled();
      expect(openToast).toHaveBeenCalledWith("Log In Failed with Invalid Email or Password.");

      expect(screen.getByTestId("login-email-input")).toHaveDisplayValue("mail@mail.com");
      expect(screen.getByTestId("login-password-input")).toHaveDisplayValue("123456");
    });

    it("should still be in login page and display toast error when login failed with other error", async () => {
      jest.mocked(signIn).mockRejectedValue({ error: "Login Other Error" } as SignInResponse);

      renderLogin();

      await userEvent.type(screen.getByTestId("login-email-input"), "mail@mail.com");
      await userEvent.type(screen.getByTestId("login-password-input"), "123456");

      await userEvent.click(screen.getByTestId("login-submit-button"));

      expect(replaceMock).not.toHaveBeenCalled();
      expect(openToast).toHaveBeenCalledWith(JSON.stringify("Login Other Error"));

      expect(screen.getByTestId("login-email-input")).toHaveDisplayValue("mail@mail.com");
      expect(screen.getByTestId("login-password-input")).toHaveDisplayValue("123456");
    });
  });
});
