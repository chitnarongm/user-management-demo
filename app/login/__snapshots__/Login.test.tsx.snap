// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container css-fyi19d-MuiGrid-root"
    data-testid="login-page-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-vok43g-MuiGrid-root"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation8 css-1qz17ui-MuiPaper-root"
      >
        <form>
          <div
            class="MuiBox-root css-11n0r2k"
            data-testid="login-page-title"
          >
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1ht8sh5-MuiTypography-root"
            >
              Log In
            </h5>
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                data-testid="login-email-label"
                for="outlined-adornment-email"
              >
                Email
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="login-email-input"
                  id="outlined-adornment-email"
                  name="email"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Email
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
                data-testid="login-email-error"
              />
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth css-zevgu-MuiFormControl-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                data-testid="login-password-label"
                for="outlined-adornment-password"
              >
                Password
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                  data-testid="login-password-input"
                  id="outlined-adornment-password"
                  name="password"
                  type="password"
                  value=""
                />
                <div
                  class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                >
                  <button
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                    data-testid="password-visibility-icon"
                    tabindex="0"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="VisibilityIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5M12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5m0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-yjsfm1"
                  >
                    <span>
                      Password
                    </span>
                  </legend>
                </fieldset>
              </div>
              <p
                class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
                data-testid="login-password-error"
              />
            </div>
            <button
              class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-colorPrimary MuiButton-fullWidth css-ejtev3-MuiButtonBase-root-MuiButton-root"
              data-testid="login-submit-button"
              tabindex="0"
              type="submit"
            >
              <span
                class="MuiButton-icon MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="LoginIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 7 9.6 8.4l2.6 2.6H2v2h10.2l-2.6 2.6L11 17l5-5zm9 12h-8v2h8c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-8v2h8z"
                  />
                </svg>
              </span>
              Log In
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
